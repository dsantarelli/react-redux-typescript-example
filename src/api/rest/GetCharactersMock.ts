import Character from '../model/Character';

const GetCharactersMock: Character[] = [{
  'name': 'Luke Skywalker',
  'height': '172',
  'mass': '77',
  'hair_color': 'blond',
  'skin_color': 'fair',
  'eye_color': 'blue',
  'birth_year': '19BBY',
  'gender': 'male',
  'homeworld': 'https://swapi.co/api/planets/1/',
  'films': [
    'https://swapi.co/api/films/2/',
  ],
  'species': [
    'https://swapi.co/api/species/1/'
  ],
  'vehicles': [
    'https://swapi.co/api/vehicles/14/',
  ],
  'starships': [
    'https://swapi.co/api/starships/12/',
  ],
  'created': '2014-12-09T13:50:51.644000Z',
  'edited': '2014-12-20T21:17:56.891000Z',
  'url': 'https://swapi.co/api/people/1/'
}];

export default GetCharactersMock;
