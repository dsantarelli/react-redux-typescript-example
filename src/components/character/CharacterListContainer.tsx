import React, { FunctionComponent } from 'react';
import CharactersListState from '../../redux/state/CharactersListState';
import ProgressBar from '../ProgressBar';
import Alert, { AlertType } from '../Alert';
import CharacterList from './CharacterList';
import Character from '../../api/model/Character';

interface Props {
    charactersList: CharactersListState,
    onCharacterSelected(character: Character): void
}
const CharacterListContainer: FunctionComponent<Props> = (props: Props) => {

    const { charactersList, onCharacterSelected } = props;

    if (charactersList.isFetching) {
        return <ProgressBar />
    }
    else {
        if (charactersList.error) {
            return <Alert type={AlertType.DANGER} message={charactersList.error} />
        }
        else {
            return charactersList.characters.length === 0 ?
                <div className="p-2">
                    <span className="font-italic">No characters found.</span>&nbsp;
                    <span role="img" aria-label="doh!">ðŸ˜“</span>
                </div> :
                <CharacterList
                    characters={charactersList.characters}
                    onCharacterSelected={(c) => onCharacterSelected(c)} />
        }
    }
}
export default CharacterListContainer;