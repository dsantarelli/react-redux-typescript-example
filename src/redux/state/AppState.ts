import CharactersListState from "./CharactersListState";
import CharacterDetailsState from "./CharacterDetailsState";

export default interface AppState {
  charactersListState: CharactersListState;
  characterDetailsState: CharacterDetailsState;
}
