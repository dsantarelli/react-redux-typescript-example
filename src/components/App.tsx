import React from 'react';
import { connect } from 'react-redux';

import AppState from '../redux/state/AppState';
import Character from '../api/model/Character';
import {
  getCharacters,
  searchCharacters
} from '../redux/actions/CharactersListActionCreators';
import CharacterList from './character/CharacterList';
import CharacterSearch from './character/CharacterSearch';
import NavigationBar from './NavigationBar';
import ProgressBar from './ProgressBar';
import Alert, { AlertType } from './Alert';


interface Props {
  getCharacters(): void,
  searchCharacters(term: string): void,
  characters: Character[],
  isFetching: boolean,
  error?: string
}
export class App extends React.Component<Props> {

  public componentDidMount() {
    if (this.props.characters.length === 0) {
      this.props.getCharacters();
    }
  }

  public render() {

    const { characters, searchCharacters, isFetching, error } = this.props;

    let content = null;
    if (isFetching) content = <ProgressBar />;
    else {
      if (error) content = <Alert type={AlertType.DANGER} message={error} />;
      else content = <CharacterList characters={characters} />;
    }

    return (
      <div className="app-container">
        <NavigationBar>
          <CharacterSearch searchCharacters={searchCharacters} />
        </NavigationBar>      
        {content}
      </div>
    );
  }
}

////////////// REDUX ///////////////

const mapStateToProps = (state: AppState) => {
  return {
    characters: state.charactersListState.characters,
    isFetching: state.charactersListState.isFetching,
    error: state.charactersListState.error
  };
};

const mapDispatchToProps = (dispatch: Function) => {
  return {
    getCharacters: () => dispatch(getCharacters()),
    searchCharacters: (term: string) => dispatch(searchCharacters(term)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
