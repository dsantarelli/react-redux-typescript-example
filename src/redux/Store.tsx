import { applyMiddleware, combineReducers, createStore, Store, AnyAction } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import AppState from './state/AppState';
import CharactersListReducer from './reducers/CharactersListReducer';
import CharacterDetailsReducer from './reducers/CharacterDetailsReducer';

const rootReducer = combineReducers<AppState>({
  charactersListState: CharactersListReducer,
  characterDetailsState: CharacterDetailsReducer
  // TODO: add sub-reducers here...
});

export default function configureStore(): Store<AppState, AnyAction> {
  return createStore(rootReducer, undefined, composeWithDevTools(applyMiddleware(thunk)));
}
