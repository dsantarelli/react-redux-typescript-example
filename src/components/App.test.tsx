import React from 'react';
import renderer, { ReactTestRenderer, ReactTestInstance } from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import GetCharactersMock from '../api/rest/GetCharactersMock';
import App from './App';
import AppState from '../redux/state/AppState';
import { Store } from 'redux';
import CharacterListContainer from './character/CharacterListContainer';
import CharacterDetailsDialog from './character/CharacterDetailsDialog';


describe('App', () => {

  const createMockStore = configureStore<AppState>([thunk]);

  const createEmptyState = (): AppState => {
    return {
      charactersListState: {
        characters: [],
        isFetching: false
      },
      characterDetailsState: {
        isFetching: false
      }
    }
  }

  const createTestRenderer = (store: Store<AppState>): ReactTestRenderer => {
    return renderer.create(
      <Provider store={store}>
        <App />
      </Provider>
    );
  }

  describe('should be rendered', () => {

    it('inside a redux Provider', () => {
      createTestRenderer(createMockStore(createEmptyState()));
    });

    it('and it should match the expected snapshot', () => {
      expect(createTestRenderer(
        createMockStore(createEmptyState()))
        .toJSON())
        .toMatchSnapshot();
    });

    describe('if a character is selected, the modal dialog should be', () => {

      const characters = GetCharactersMock;
      const state = createEmptyState();
      state.charactersListState.characters = characters;
      let component: ReactTestRenderer;
      let characterListContainer: ReactTestInstance;
      let selectedCharacterDialog: ReactTestInstance;

      beforeEach(() => {
        component = createTestRenderer(createMockStore(state));
        characterListContainer = component.root.findByType(CharacterListContainer);
        selectedCharacterDialog = component.root.findByType(CharacterDetailsDialog);
      })

      it('opened', () => {
        expect(selectedCharacterDialog.props.isOpen).toBe(false);
        characterListContainer.props.onCharacterSelected(characters[0]);
        expect(selectedCharacterDialog.props.isOpen).toBe(true);
      });

      it('and then it can be closed', () => {
        expect(selectedCharacterDialog.props.isOpen).toBe(false);
        characterListContainer.props.onCharacterSelected(characters[0]);
        expect(selectedCharacterDialog.props.isOpen).toBe(true);
        selectedCharacterDialog.props.onClose();
        expect(selectedCharacterDialog.props.isOpen).toBe(false);
      });
    });
  });
});