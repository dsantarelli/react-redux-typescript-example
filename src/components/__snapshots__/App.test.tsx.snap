// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should be rendered and it should match the expected snapshot 1`] = `
<div
  className="app-container"
>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <span
      className="navbar-brand mb-0 h1"
    >
      Star Wars Characters
    </span>
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <input
        aria-label="Search"
        className="form-control mr-sm-2"
        onChange={[Function]}
        placeholder="Search"
        type="search"
        value=""
      />
      <button
        className="btn btn-secondary my-2 my-sm-0"
        type="submit"
      >
        Search
      </button>
    </form>
  </nav>
  <div
    className="p-2"
  >
    <span
      className="font-italic"
    >
      No characters found.
    </span>
    Â 
    <span
      aria-label="doh!"
      role="img"
    >
      ðŸ˜“
    </span>
  </div>
  <div>
    <div
      aria-hidden={true}
      className="modal fade"
      onClick={[Function]}
      role="dialog"
      style={
        Object {
          "display": "none",
          "zIndex": 1041,
        }
      }
      tabIndex="-1"
    >
      <div
        className="modal-dialog modal-dialog modal-dialog-centered modal-lg"
        onClick={[Function]}
        role="document"
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-header bg-dark text-white"
          >
            <h5
              className="modal-title"
            >
              
            </h5>
            <button
              aria-label="Close"
              className="close text-white"
              data-dismiss="modal"
              onClick={[Function]}
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
          </div>
          <div
            className="modal-body"
          >
            
          </div>
          <div
            className="modal-footer"
          >
            <button
              className="btn btn-secondary"
              onClick={[Function]}
              type="button"
            >
              Close
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`App should be rendered and match the expected snapshot 1`] = `
<div
  className="app-container"
>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <span
      className="navbar-brand mb-0 h1"
    >
      Star Wars Characters
    </span>
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <input
        aria-label="Search"
        className="form-control mr-sm-2"
        onChange={[Function]}
        placeholder="Search"
        type="search"
        value=""
      />
      <button
        className="btn btn-secondary my-2 my-sm-0"
        type="submit"
      >
        Search
      </button>
    </form>
  </nav>
  <div
    className="p-2"
  >
    <span
      className="font-italic"
    >
      No characters found.
    </span>
    Â 
    <span
      aria-label="doh!"
      role="img"
    >
      ðŸ˜“
    </span>
  </div>
  <div>
    <div
      aria-hidden={true}
      className="modal fade"
      onClick={[Function]}
      role="dialog"
      style={
        Object {
          "display": "none",
          "zIndex": 1041,
        }
      }
      tabIndex="-1"
    >
      <div
        className="modal-dialog modal-dialog modal-dialog-centered modal-lg"
        onClick={[Function]}
        role="document"
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-header bg-dark text-white"
          >
            <h5
              className="modal-title"
            >
              
            </h5>
            <button
              aria-label="Close"
              className="close text-white"
              data-dismiss="modal"
              onClick={[Function]}
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
          </div>
          <div
            className="modal-body"
          >
            
          </div>
          <div
            className="modal-footer"
          >
            <button
              className="btn btn-secondary"
              onClick={[Function]}
              type="button"
            >
              Close
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`App should be rendered with the expected snapshot 1`] = `
<div
  className="app-container"
>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <span
      className="navbar-brand mb-0 h1"
    >
      Star Wars Characters
    </span>
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <input
        aria-label="Search"
        className="form-control mr-sm-2"
        onChange={[Function]}
        placeholder="Search"
        type="search"
        value=""
      />
      <button
        className="btn btn-secondary my-2 my-sm-0"
        type="submit"
      >
        Search
      </button>
    </form>
  </nav>
  <div
    className="p-2"
  >
    <span
      className="font-italic"
    >
      No characters found.
    </span>
    Â 
    <span
      aria-label="doh!"
      role="img"
    >
      ðŸ˜“
    </span>
  </div>
  <div>
    <div
      aria-hidden={true}
      className="modal fade"
      onClick={[Function]}
      role="dialog"
      style={
        Object {
          "display": "none",
          "zIndex": 1041,
        }
      }
      tabIndex="-1"
    >
      <div
        className="modal-dialog modal-dialog modal-dialog-centered modal-lg"
        onClick={[Function]}
        role="document"
      >
        <div
          className="modal-content"
        >
          <div
            className="modal-header bg-dark text-white"
          >
            <h5
              className="modal-title"
            >
              
            </h5>
            <button
              aria-label="Close"
              className="close text-white"
              data-dismiss="modal"
              onClick={[Function]}
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
          </div>
          <div
            className="modal-body"
          >
            
          </div>
          <div
            className="modal-footer"
          >
            <button
              className="btn btn-secondary"
              onClick={[Function]}
              type="button"
            >
              Close
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
